tpm2/src/util/ski.c

```diff
@@ -0,0 +1,560 @@
+                                    &wk_pub_data,
+                                    &wk_pub_size,
+                                    wk_pub_offset,
+                                    &input.wk_priv,
+                                    &wk_priv_data,
+                                    &wk_priv_size, wk_priv_offset))
+  {
+    kmyth_log(LOG_ERR, "unable to marshal data for ski file ... exiting");
+    free(pcr_select_data);
+    free(sk_pub_data);
+    free(sk_priv_data);
+    free(wk_pub_data);
+    free(wk_priv_data);
+    return 1;
+  }
+
+  // validate that all data to be written is non-NULL and non-empty
+  if (pcr_select_data == NULL ||
+      pcr_select_size == 0 ||
+      sk_pub_data == NULL ||
+      sk_pub_size == 0 ||
+      sk_priv_data == NULL ||
+      sk_priv_size == 0 ||
+      wk_pub_data == NULL ||
+      wk_pub_size == 0 ||
+      wk_priv_data == NULL ||
+      wk_priv_size == 0 ||
+      input.cipher.cipher_name == NULL ||
+      strlen(input.cipher.cipher_name) == 0 ||
+      input.enc_data == NULL || input.enc_data_size == 0)
+  {
+    kmyth_log(LOG_ERR, "cannot write empty sections ... exiting");
+    free(pcr_select_data);
+    free(sk_pub_data);
+    free(sk_priv_data);
+    free(wk_pub_data);
+    free(wk_priv_data);
+    return 1;
+  }
+
+//Encode each portion of the file in base64
+  uint8_t *pcr64_select_data = NULL;
+  size_t pcr64_select_size = 0;
+  uint8_t *sk64_pub_data = NULL;
+  size_t sk64_pub_size = 0;
+  uint8_t *sk64_priv_data = NULL;
+  size_t sk64_priv_size = 0;
+  uint8_t *wk64_pub_data = NULL;
+  size_t wk64_pub_size = 0;
+  uint8_t *wk64_priv_data = NULL;
+  size_t wk64_priv_size = 0;
+  uint8_t *enc64_data = NULL;
+  size_t enc64_data_size = 0;
+
+  if (encodeBase64Data
+      (pcr_select_data, pcr_select_size, &pcr64_select_data, &pcr64_select_size)
+      || encodeBase64Data(sk_pub_data, sk_pub_size, &sk64_pub_data,
+                          &sk64_pub_size)
+      || encodeBase64Data(sk_priv_data, sk_priv_size, &sk64_priv_data,
+                          &sk64_priv_size)
+      || encodeBase64Data(wk_pub_data, wk_pub_size, &wk64_pub_data,
+                          &wk64_pub_size)
+      || encodeBase64Data(wk_priv_data, wk_priv_size, &wk64_priv_data,
+                          &wk64_priv_size)
+      || encodeBase64Data(input.enc_data, input.enc_data_size, &enc64_data,
+                          &enc64_data_size))
+  {
+    kmyth_log(LOG_ERR, "error base64 encoding ski string ... exiting");
+    free(pcr_select_data);
+    free(sk_pub_data);
+    free(sk_priv_data);
+    free(wk_pub_data);
+    free(wk_priv_data);
+    free(pcr64_select_data);
+    free(sk64_pub_data);
+    free(sk64_priv_data);
+    free(wk64_pub_data);
+    free(wk64_priv_data);
+    free(enc64_data);
+    return 1;
+  }
+
+  free(pcr_select_data);
+  free(sk_pub_data);
+  free(sk_priv_data);
+  free(wk_pub_data);
+  free(wk_priv_data);
+
+  //At this point the data is all formatted, it's time to create the string
```

Comment:
Since we already have the `Ski` structure in hand (and the encoded data) why aren't you allocating all the necessary memory up front?

Reply:
We could, but I think concat is pretty easy to read and more flexible as implemented.